pkg/jtracer/jtracer.go

```diff
@@ -14,16 +14,109 @@
 
 package jtracer
 
-import "github.com/opentracing/opentracing-go"
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"go.opentelemetry.io/otel"
+	otbridge "go.opentelemetry.io/otel/bridge/opentracing"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
 
 type JTracer struct {
-	OT opentracing.Tracer
+	OT   opentracing.Tracer
+	OTEL *sdktrace.TracerProvider
+	// logger *zap.Logger
 }
 
-func OT(t opentracing.Tracer) JTracer {
-	return JTracer{OT: t}
+var once sync.Once
+
+func New() JTracer {
+	opentracingTracer, otelTracerProvider := initBoth()
+
+	// Shutdown the tracerProvider to clean up resources
+	defer func() {
+		if err := otelTracerProvider.Shutdown(context.Background()); err != nil {
+			log.Fatal("Error shutting down tracer provider: %w", err)
+		}
+	}()
+	return JTracer{OT: opentracingTracer, OTEL: otelTracerProvider}
 }
 
 func NoOp() JTracer {
-	return JTracer{OT: opentracing.NoopTracer{}}
+	return JTracer{OT: opentracing.NoopTracer{}, OTEL: &sdktrace.TracerProvider{}}
+}
+
+// initBoth initializes OpenTelemetry SDK and uses OTel-OpenTracing Bridge
+func initBoth() (opentracing.Tracer, *sdktrace.TracerProvider) {
+	ctx := context.Background()
+	traceExporter, err := newExporter(ctx)
+	if err != nil {
+		log.Fatal("failed to create exporter", zap.Any("error", err))
+	}
+
+	// Register the trace exporter with a TracerProvider, using a batch
+	// span processor to aggregate spans before export.
+	bsp := sdktrace.NewBatchSpanProcessor(traceExporter)
+	tracerProvider := sdktrace.NewTracerProvider(
+		sdktrace.WithSpanProcessor(bsp),
+		sdktrace.WithResource(resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String("otlp"),
+		)),
+	)
+
+	once.Do(func() {
+		otel.SetTextMapPropagator(
+			propagation.NewCompositeTextMapPropagator(
+				propagation.TraceContext{},
+				propagation.Baggage{},
+			))
+	})
+
+	// Use the bridgeTracer as your OpenTracing tracer(otTrace).
+	otTracer, _ := otbridge.NewTracerPair(tracerProvider.Tracer(""))
+
+	return otTracer, tracerProvider
+}
+
+// newExporter returns a console exporter.
+func newExporter(ctx context.Context) (sdktrace.SpanExporter, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	server := grpc.NewServer()
+
+	lis, _ := net.Listen("tcp", ":0")
+	go func() {
+		err := server.Serve(lis)
+		log.Fatal("failed to create gRPC server connection: %w", err)
```

Comment:
@yurishkuro how would I log this err? you told something like wrapping this into upstream logger.

Reply:
why are you starting a server from `newExporter`?

Reply:
https://github.com/jaegertracing/jaeger/blob/e40a3e5a0ea1ab2a4634348bc3dd2093775651af/cmd/tracegen/main.go#L102

Reply:
That's my bad. I would consider this otlp-grpc https://github.com/jaegertracing/jaeger/blob/e40a3e5a0ea1ab2a4634348bc3dd2093775651af/cmd/tracegen/main.go#L115-L117
Right?

Reply:
yes